include/nbl/builtin/hlsl/bxdf/ndf/ggx.hlsl

```diff
@@ -0,0 +1,286 @@
+            const scalar_type VdotH_etaLdotH = (cache.getVdotH() + query.getOrientedEta() * cache.getLdotH());
+            // VdotHLdotH is negative under transmission, so this factor is negative
+            factor *= -scalar_type(2.0) * cache.getVdotHLdotH() / (VdotH_etaLdotH * VdotH_etaLdotH);
+        }
+        return NG * factor;
+    }
+
+    scalar_type devsh_part(scalar_type NdotX2)
+    {
+        assert(a2 >= numeric_limits<scalar_type>::min);
+        return sqrt(a2 + one_minus_a2 * NdotX2);
+    }
+
+    scalar_type G1_wo_numerator(scalar_type absNdotX, scalar_type NdotX2)
+    {
+        return scalar_type(1.0) / (absNdotX + devsh_part(NdotX2));
+    }
+
+    scalar_type G1_wo_numerator_devsh_part(scalar_type absNdotX, scalar_type devsh_part)
+    {
+        // numerator is 2 * NdotX
+        return scalar_type(1.0) / (absNdotX + devsh_part);
+    }
+
+    // without numerator, numerator is 2 * NdotV * NdotL, we factor out 4 * NdotV * NdotL, hence 0.5
+    template<class Query, class LS, class Interaction NBL_FUNC_REQUIRES(ggx_concepts::G2XQuery<Query> && LightSample<LS> && surface_interactions::Isotropic<Interaction>)
+    scalar_type correlated(NBL_CONST_REF_ARG(Query) query, NBL_CONST_REF_ARG(LS) _sample, NBL_CONST_REF_ARG(Interaction) interaction)
+    {
+        BxDFClampMode _clamp = query.getClampMode();
+        assert(_clamp != BxDFClampMode::BCM_NONE);
+
+        scalar_type Vterm = _sample.getNdotL(_clamp) * query.getDevshV();
+        scalar_type Lterm = interaction.getNdotV(_clamp) * query.getDevshL();
+        return scalar_type(0.5) / (Vterm + Lterm);
+    }
+
+    template<class Query, class LS, class Interaction, class MicrofacetCache NBL_FUNC_REQUIRES(ggx_concepts::G2XQuery<Query> && LightSample<LS> && surface_interactions::Isotropic<Interaction> && ReadableIsotropicMicrofacetCache<MicrofacetCache>)
+    scalar_type G2_over_G1(NBL_CONST_REF_ARG(Query) query, NBL_CONST_REF_ARG(LS) _sample, NBL_CONST_REF_ARG(Interaction) interaction, NBL_CONST_REF_ARG(MicrofacetCache) cache)
+    {
+        BxDFClampMode _clamp = query.getClampMode();
+        assert(_clamp != BxDFClampMode::BCM_NONE);
+
+        scalar_type G2_over_G1;
+        scalar_type NdotV = interaction.getNdotV(_clamp);
+        scalar_type NdotL = _sample.getNdotL(_clamp);
+        scalar_type devsh_v = query.getDevshV();
+        scalar_type devsh_l = query.getDevshL();
+        if (cache.isTransmission())
+        {
+            if (NdotV < 1e-7 || NdotL < 1e-7)
+                return 0.0;
+            scalar_type onePlusLambda_V = scalar_type(0.5) * (devsh_v / NdotV + scalar_type(1.0));
+            scalar_type onePlusLambda_L = scalar_type(0.5) * (devsh_l / NdotL + scalar_type(1.0));
+            G2_over_G1 = bxdf::beta<scalar_type>(onePlusLambda_L, onePlusLambda_V) * onePlusLambda_V;
+        }
+        else
+        {
+            G2_over_G1 = NdotL * (devsh_v + NdotV); // alternative `Vterm+NdotL*NdotV /// NdotL*NdotV could come as a parameter
```

Comment:
remove the comment, you already explained why clamping would make it impossible to use `NdotLNdotV` as parameter